lib/ExpensiMark.js

```diff
@@ -628,4 +612,25 @@ export default class ExpensiMark {
         }
         return replacedText;
     }
+
+    /**
+     * Handle whitespace for the content of blockquote if the text matches the regex or else just return the original text
+     *
+     * @param {RegExp} regex
+     * @param {String} textToCheck
+     * @param {Function} replacement
+     *
+     * @returns {String}
+     */
+    formatWhitespaceForQuote(regex, textToCheck, replacement) {
```

Comment:
Can you explain why this change is required? What would be wrong if we just change the regex and keep everything as is?

Reply:
The regex is only used to check if the input text contains quotes, see [here](https://github.com/Expensify/expensify-common/blob/be625f55058111793bbcfe3f64788ea0a8f9d705/lib/ExpensiMark.js#L555-L561). It'll skip to translate input '> \n>', but will still create empty `blockquote` tags for input text like
```
> a
test
> 
> 
```
See below videos for comparison
**Before** 

https://user-images.githubusercontent.com/117511920/233652801-26207fb7-ce1c-4611-9739-e2b82c09e86f.mov

**After** 

https://user-images.githubusercontent.com/117511920/233652982-69214585-ee91-4ba0-b739-91c8bc39023b.mov


I think we need to avoid creating all empty `blockquote` tags, so I made this change to ensure consistency. Do you agree with this?

Reply:
Right, that would be great. But can't we do it by just altering the logic here https://github.com/Expensify/expensify-common/blob/be625f55058111793bbcfe3f64788ea0a8f9d705/lib/ExpensiMark.js#L574-L589. And without the use of an additional regex :sweat_smile:.

Reply:
@s77rt Thanks for the comment.

It should be feasible to achieve this by modifying method `modifyTextForQuote` technically but I feel it's tricky to handle 
1. blank line within a multiple line quote. Note that we only keep one blank line inside a quote between two rows.
2. trim spaces for each row of a quote
3. retain original text for empty quote, like '> \n> '

We need to find a way to retain original text, like for case '> \n> ', but the current implementation pollute `textSplitted[i]` directly and depends on the value of `textSplitted[I]` to handle blank row inside a quote.
So I think it'll be better to let `textToFormat` save the original text and extract the logic to format a quote into another method. Then both method  `modifyTextForQuote` and `formatWhitespaceForQuote` are more readable.

As for the regex, I think it's also necessary to improve it to make it consistent with the implementation of method `formatWhitespaceForQuote`

Reply:
@eh2077 Thanks for the explanation. I will have to check again to see if we can improve that. Maybe we can relay on the next line `textSplitted[i + 1]` to decide whether the current `>` should be kept as is or considered as a blockquote.

Reply:
@eh2077 I think I managed to get this working without regex. The key is to defer parsing the "tricky" lines till we manage the easy ones so it become clear if those "tricky" lines should be parsed as part of the quote or not. Do you want to work on top of that? Also can you please add more tests?

<details>
<summary>diff</summary>

```diff
diff --git a/lib/ExpensiMark.js b/lib/ExpensiMark.js
index bac4b81..21a4c2a 100644
--- a/lib/ExpensiMark.js
+++ b/lib/ExpensiMark.js
@@ -155,7 +155,7 @@ export default class ExpensiMark {
                 // inline code blocks. A single prepending space should be stripped if it exists
                 process: (textToProcess, replacement) => {
                     const regex = new RegExp(
-                        /\n?^&gt; ?(?![^<]*(?:<\/pre>|<\/code>))([^\v\n\r]+)\n?/gm,
+                        /\n?^&gt; *(?! )(?![^<]*(?:<\/pre>|<\/code>))([^\v\n\r]+)\n?/gm,
                     );
                     return this.modifyTextForQuote(regex, textToProcess, replacement);
                 },
@@ -550,6 +550,7 @@ export default class ExpensiMark {
     modifyTextForQuote(regex, textToCheck, replacement) {
         let replacedText = '';
         let textToFormat = '';
+        let maybeLinesContainsTextToFormat = [];
         const match = textToCheck.match(regex);
 
         // If there's matches we need to modify the quotes
@@ -565,22 +566,17 @@ export default class ExpensiMark {
                     insideCodefence = Str.contains(textSplitted[i], '<pre>');
                 }
 
-                // We only want to modify lines starting with &gt; that is not codefence
                 if (Str.startsWith(textSplitted[i], '&gt;') && !insideCodefence) {
-                    // Avoid blank lines starting with &gt;
-                    if (textSplitted[i].trim() !== '&gt;') {
-                        textSplitted[i] = textSplitted[i].substr(4).trim();
-                        textSplitted[i] = textSplitted[i].trim();
-                        textToFormat += `${textSplitted[i]}\n`;
-
-                        // We need ensure that index i-1 >= 0
-                    } else if (i > 0) {
-                        // Certificate that we will have only one blank row
-                        if (textSplitted[i - 1].trim() !== '') {
-                            textSplitted[i] = textSplitted[i].substr(4).trim();
-                            textSplitted[i] = textSplitted[i].trim();
-                            textToFormat += `${textSplitted[i]}\n`;
+                    if (textToFormat !== '' || textSplitted[i].trim() !== '&gt;') {
+                        if (maybeLinesContainsTextToFormat.length) {
+                            maybeLinesContainsTextToFormat.forEach(j => {
+                                textToFormat += `${textSplitted[j].substr(4).trim()}\n`;
+                            });
+                            maybeLinesContainsTextToFormat = [];
                         }
+                        textToFormat += `${textSplitted[i].substr(4).trim()}\n`;
+                    } else {
+                        maybeLinesContainsTextToFormat.push(i);
                     }
                 } else {
                     // Make sure we will only modify if we have Text needed to be formatted for quote
@@ -588,6 +584,11 @@ export default class ExpensiMark {
                         textToFormat = textToFormat.replace(/\n$/, '');
                         replacedText += replacement(textToFormat);
                         textToFormat = '';
+                    } else if (maybeLinesContainsTextToFormat.length) {
+                        maybeLinesContainsTextToFormat.forEach(j => {
+                            replacedText += `${textSplitted[j]}\n`;
+                        });
+                        maybeLinesContainsTextToFormat = [];
                     }
 
                     // We dont want a \n after the textSplitted if it is the last row
@@ -608,6 +609,16 @@ export default class ExpensiMark {
             if (textToFormat !== '') {
                 textToFormat = textToFormat.replace(/\n$/, '');
                 replacedText += replacement(textToFormat);
+                textToFormat = '';
+            } else if (maybeLinesContainsTextToFormat.length) {
+                maybeLinesContainsTextToFormat.forEach(j => {
+                    if (j === textSplitted.length - 1) {
+                        replacedText += `${textSplitted[j]}`;
+                    } else {
+                        replacedText += `${textSplitted[j]}\n`;
+                    }
+                });
+                maybeLinesContainsTextToFormat = [];
             }
 
             // Replace all the blank lines between quotes, if there are only blank lines between them
```

</details>

cc @yuwenmemon

Reply:
@s77rt I like the idea to defer parsing quote lines and actually what I did in the PR uses the same idea but it defers parsing all lines start with '>' instead of only empty lines. I tested the diff you attached and it almost worked. I just made one additional change to replace this line
```js
textToFormat = textToFormat.replace(/\n$/, '');
```
to 
```js
textToFormat = textToFormat.trim().replace(/\n{2,}/g, '\n\n');
```
and then we can handle line breaks correctly. Below is the updated diff 

<details>
  <summary>Updated diff</summary>
  
```diff
diff --git a/lib/ExpensiMark.js b/lib/ExpensiMark.js
index 9f69183..f5e1239 100644
--- a/lib/ExpensiMark.js
+++ b/lib/ExpensiMark.js
@@ -163,7 +163,7 @@ export default class ExpensiMark {
                 // inline code blocks. A single prepending space should be stripped if it exists
                 process: (textToProcess, replacement) => {
                     const regex = new RegExp(
-                        /\n?^&gt; ?(?![^<]*(?:<\/pre>|<\/code>))([^\v\n\r]+)\n?/gm,
+                        /\n?^&gt; *(?! )(?![^<]*(?:<\/pre>|<\/code>))([^\v\n\r]+)\n?/gm,
                     );
                     return this.modifyTextForQuote(regex, textToProcess, replacement);
                 },
@@ -560,6 +560,7 @@ export default class ExpensiMark {
     modifyTextForQuote(regex, textToCheck, replacement) {
         let replacedText = '';
         let textToFormat = '';
+        let maybeLinesContainsTextToFormat = [];
         const match = textToCheck.match(regex);
 
         // If there's matches we need to modify the quotes
@@ -575,29 +576,29 @@ export default class ExpensiMark {
                     insideCodefence = Str.contains(textSplitted[i], '<pre>');
                 }
 
-                // We only want to modify lines starting with &gt; that is not codefence
                 if (Str.startsWith(textSplitted[i], '&gt;') && !insideCodefence) {
-                    // Avoid blank lines starting with &gt;
-                    if (textSplitted[i].trim() !== '&gt;') {
-                        textSplitted[i] = textSplitted[i].substr(4).trim();
-                        textSplitted[i] = textSplitted[i].trim();
-                        textToFormat += `${textSplitted[i]}\n`;
-
-                        // We need ensure that index i-1 >= 0
-                    } else if (i > 0) {
-                        // Certificate that we will have only one blank row
-                        if (textSplitted[i - 1].trim() !== '') {
-                            textSplitted[i] = textSplitted[i].substr(4).trim();
-                            textSplitted[i] = textSplitted[i].trim();
-                            textToFormat += `${textSplitted[i]}\n`;
+                    if (textToFormat !== '' || textSplitted[i].trim() !== '&gt;') {
+                        if (maybeLinesContainsTextToFormat.length) {
+                            maybeLinesContainsTextToFormat.forEach(j => {
+                                textToFormat += `${textSplitted[j].substr(4).trim()}\n`;
+                            });
+                            maybeLinesContainsTextToFormat = [];
                         }
+                        textToFormat += `${textSplitted[i].substr(4).trim()}\n`;
+                    } else {
+                        maybeLinesContainsTextToFormat.push(i);
                     }
                 } else {
                     // Make sure we will only modify if we have Text needed to be formatted for quote
                     if (textToFormat !== '') {
-                        textToFormat = textToFormat.replace(/\n$/, '');
+                        textToFormat = textToFormat.trim().replace(/\n{2,}/g, '\n\n');
                         replacedText += replacement(textToFormat);
                         textToFormat = '';
+                    } else if (maybeLinesContainsTextToFormat.length) {
+                        maybeLinesContainsTextToFormat.forEach(j => {
+                            replacedText += `${textSplitted[j]}\n`;
+                        });
+                        maybeLinesContainsTextToFormat = [];
                     }
 
                     // We dont want a \n after the textSplitted if it is the last row
@@ -616,8 +617,18 @@ export default class ExpensiMark {
 
             // When loop ends we need the last quote to be formatted if we have quotes at last rows
             if (textToFormat !== '') {
-                textToFormat = textToFormat.replace(/\n$/, '');
+                textToFormat = textToFormat.trim().replace(/\n{2,}/g, '\n\n');
                 replacedText += replacement(textToFormat);
+                textToFormat = '';
+            } else if (maybeLinesContainsTextToFormat.length) {
+                maybeLinesContainsTextToFormat.forEach(j => {
+                    if (j === textSplitted.length - 1) {
+                        replacedText += `${textSplitted[j]}`;
+                    } else {
+                        replacedText += `${textSplitted[j]}\n`;
+                    }
+                });
+                maybeLinesContainsTextToFormat = [];
             }
 
             // Replace all the blank lines between quotes, if there are only blank lines between them
```
</details>

Based on the updated diff above, I think extracting the logic to format a `blockquote` tag into another method does improve the readability of the code. Do you agree with this? Or I'd like to hear your concerns about refactoring it into a new method.

Below are the test cases I tested and will be added
<details>
  <summary>Test cases</summary>

```js
test('Test quotes markdown replacement skipping blank quotes ', () => {
    const testString = '> \n>';
    const resultString = '&gt; <br />&gt;';
    expect(parser.replace(testString)).toBe(resultString);
});

test('Test quotes markdown replacement with text starts with blank quote', () => {
    const testString = '> \ntest';
    const resultString = '&gt; <br />test';
    expect(parser.replace(testString)).toBe(resultString);
});

test('Test quotes markdown replacement with quotes starts with blank quote row', () => {
    const testString = '> \n>test';
    const resultString = '<blockquote>test</blockquote>';
    expect(parser.replace(testString)).toBe(resultString);
});

test('Test quotes markdown replacement with quotes ends with blank quote rows', () => {
    const testString = '>test\n> \n>';
    const resultString = '<blockquote>test</blockquote>';
    expect(parser.replace(testString)).toBe(resultString);
});

test('Test quotes markdown replacement with quotes includes a middle blank quote row', () => {
    const testString = '>test\n> \n>test';
    const resultString = '<blockquote>test<br /><br />test</blockquote>';
    expect(parser.replace(testString)).toBe(resultString);
});

test('Test quotes markdown replacement with quotes includes multiple middle blank quote rows', () => {
    const testString = '>test\n> \n> \n>test';
    const resultString = '<blockquote>test<br /><br />test</blockquote>';
    expect(parser.replace(testString)).toBe(resultString);
});

test('Test quotes markdown replacement with text includes blank quotes', () => {
    const testString = '> \n>quote1 line a\n> quote1 line b\ntest\n> \ntest\n>quote2 line a\n> \n> \n>quote2 line b with an empty line above';
    const resultString = '<blockquote>quote1 line a<br />quote1 line b</blockquote>test<br />&gt; <br />test<br /><blockquote>quote2 line a<br /><br />quote2 line b with an empty line above</blockquote>';
    expect(parser.replace(testString)).toBe(resultString);
});
```
  
</details>

Reply:
I noticed the diff in line breaks but I don't understand why we are doing it that way. The current implementation is buggy and whatever is done with line breaks seems unintended. It appears to me that the new behaviour of line breaks makes more sense and it's something that would match your expectations. Or maybe I'm missing something here?

___


Current:
```
> a
>
> b
```
One line break

```
> a
>
>
> b
```
One line breaks

```
> a
>
>
>
> b
```
Two line breaks

___


New:
```
> a
>
> b
```
One line break

```
> a
>
>
> b
```
Two line breaks

```
> a
>
>
>
> b
```
Three line breaks

Reply:
@s77rt @yuwenmemon 
I tested following quotes and attached the screenshots FYI. Please let me know if you're fine with keeping one blank row between two rows of a quote.

```
> quote 1
> 
> after 1 blank row

> quote 2
> 
> 
> after 2 blank rows

> quote 3
> 
> 
> 
> after 3 blank rows

> quote 4
> 
> 
> 
> 
> after 4 blank rows

> quote 5
> 
> 
> 
> 
> 
> after 5 blank rows
```

| new.expensify.com  | This PR |
| ------------- | ------------- |
| <img width="190" alt="Screenshot 2023-04-24 at 4 24 35 PM" src="https://user-images.githubusercontent.com/117511920/233941016-a374ffc7-a036-472f-8631-4926b3267b51.png">  | <img width="190" alt="Screenshot 2023-04-24 at 4 24 02 PM" src="https://user-images.githubusercontent.com/117511920/233941002-16a94d37-ce35-4b40-8617-faeede6c0d11.png"> |

Reply:
I think we should ask on Slack. Can you please compile 3 screenshots
1. The pair line break (Staging Expensify)
2. The one line break
3. The all line break